kafka-client/src/main/java/dev/responsive/kafka/store/CommitBuffer.java

```diff
@@ -242,6 +287,78 @@ private AtomicWriteResult flush(final long offset, final UUID txnid) {
     );
     buffer.clear();
 
+    maybeTruncateChangelog(offset);
+    return AtomicWriteResult.success(partition);
+  }
+
+  private AtomicWriteResult drain(
+      final Collection<AtomicWriter<K>> writers,
+      final Function<AtomicWriter<K>, CompletionStage<AtomicWriteResult>> action
+  ) {
+    return drain(
+        writers,
+        action,
+        AtomicWriter::partition,
+        partition,
+        ResponsiveConfig.MAX_CONCURRENT_REMOTE_WRITES_DEFAULT
+    );
+  }
+
+  // Visible for Testing
+  static <I> AtomicWriteResult drain(
+      final Collection<I> writers,
+      final Function<I, CompletionStage<AtomicWriteResult>> action,
+      final Function<I, Integer> keyFunction,
+      final int partition,
+      final int maxConcurrentWrites
+  ) {
+    final var results = new HashMap<Integer, CompletableFuture<AtomicWriteResult>>();
+    try {
+      final AtomicReference<AtomicWriteResult> failed = new AtomicReference<>(null);
+
+      for (final I writer : writers) {
+        if (results.size() >= maxConcurrentWrites) {
+          CompletableFuture.anyOf(results.values().toArray(CompletableFuture[]::new)).get();
+          results.values().removeIf(CompletableFuture::isDone);
+        }
+
+        if (failed.get() != null) {
+          return failed.get();
+        }
+
+        results.put(
+            keyFunction.apply(writer),
+            action.apply(writer).thenApply(r -> maybeSetFailure(failed, r)).toCompletableFuture());
+      }
+
+      // drain any remaining
+      for (final var fut : results.values()) {
```

Comment:
sorry for so many nit comments, but `var fut` is very cryptic 🙃

Reply:
haha fut = future? is that not standard? I can change that